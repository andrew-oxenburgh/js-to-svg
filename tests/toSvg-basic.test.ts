'use strict'
import {expect} from '@jest/globals'

const toSvg = require('../src/index')
const pd = require('pretty-data').pd
const fs = require('fs')

function expectSnapshot(json: object) {
   const wrapped = toSvg.createSvgObject(
      {width: 1000, height: 1000},
      [json]
   )
   let actual = toSvg.stringify(wrapped)
   actual = pd.xml(actual)
   const testname = expect.getState().currentTestName
   const dir = './tests/__out__/'
   const filename = dir + testname.replace(/ /g, '_') + '.html'

   !fs.existsSync(dir) && fs.mkdirSync(dir)
   fs.existsSync(filename) && fs.unlinkSync(filename)
   fs.writeFileSync(filename, actual)

   expect(actual).toMatchSnapshot()
}

describe('clipPath', () => {
   test('heart', () => {
      const json = [
         toSvg.elem('clipPath',
            {id: 'myClip'},
            [
               toSvg.circle({cx: 40, cy: 40, r: 20, fill:'red'})
            ]
         ),
         toSvg.path({
            id: 'heart',
            d: 'M10,30 A20,20,0,0,1,50,30 A20,20,0,0,1,90,30 Q90,60,50,90 Q10,60,10,30 Z'
         }),
         toSvg.elem('use',
            {
               'clip-path': 'url(#myClip)',
               href: '#heart',
               fill: 'red'
            })
      ]
      expectSnapshot(json)
   })
})
describe('animation', () => {
   describe('rect', () => {
      test('rx', () => {
         const size = {width: 200, height: 200}
         const json = toSvg.rect({
            origin: {x: 0, y: 0},
            size,
            fill: 'red',
            stroke: 'none',
         }, [
            toSvg.animate({
               attributeName: 'rx',
               values: '0;200;0',
               dur: '2s',
               repeatCount: 'indefinite'
            }
            )
         ])
         expectSnapshot(json)
      })
   })
   describe('circle', () => {
      test('radius', () => {
         const json = [
            toSvg.rect({
               rect: {x: 0, y: 0, width: 2000, height: 2000},
               fill: 'red'
            }),
            toSvg.circle({
               center: {x: 500, y: 500},
               r: 50,
               fill: 'red',
               stroke: 'none',
            }, [
               toSvg.animate({
                  attributeName: 'r',
                  values: '0;100;0',
                  dur: '5s',
                  repeatCount: 'indefinite'
               }
               ),
               toSvg.animate({
                  attributeName: 'fill',
                  values: 'blue;red;green',
                  dur: '3s',
                  repeatCount: 'indefinite'
               }
               ),
               toSvg.animate({
                  attributeName: 'cx',
                  values: '0;400',
                  dur: '5s',
                  repeatCount: 'indefinite'
               }
               ),
               toSvg.animate({
                  attributeName: 'cy',
                  values: '0;400',
                  dur: '5s',
                  repeatCount: 'indefinite'
               }
               )
            ])]
         expectSnapshot(json)
      })
   })
})
describe('svg', () => {
   describe('rect', () => {
      test('standard', () => {
         const json = toSvg.rect({
            width: 777,
            height: 888,
            x: 0,
            y: 1,
            fill: 'red',
            stroke: 'none'
         })
         expectSnapshot(json)
      })
      test('origin and size', () => {
         const origin = {x: 0, y: 1}
         const size = {width: 345, height: 678}
         const json = toSvg.rect({
            origin,
            size,
            fill: 'red',
            stroke: 'none'
         })
         expectSnapshot(json)
      })
      test('rect', () => {
         const rect = {x: 1, y: 2, width: 100, height: 200}
         const json = toSvg.rect({
            rect,
            fill: 'red',
            stroke: 'none'
         })
         expectSnapshot(json)
      })
   })
   describe('circle', () => {
      test('standard', () => {
         const json = toSvg.circle({
            cx: 500,
            cy: 500,
            r: 300,
            fill: 'yellow',
            stroke: 'none'
         })
         expectSnapshot(json)
      })
      test('point and radius', () => {
         const center = {x: 181, y: 191}
         const json = toSvg.circle({
            center,
            r: 233,
            fill: 'red',
            stroke: 'none'
         })
         expectSnapshot(json)
      })
      test('bounding square', () => {
         const square = {x: 0, y: 0, side: 500}
         const json = toSvg.circle({
            square,
            fill: 'red',
            stroke: 'none'
         })
         expectSnapshot(json)
      })
   })

   describe('doc', () => {
      test('empty doc', () => {
         expectSnapshot([])
      })
   })
   describe('path', () => {
      test('a short line', () => {
         const json = toSvg.path({
            d: 'M 100 100 H 300 Z',
            fill: 'red',
            stroke: 'orange',
            'stroke-width': 10
         })
         expectSnapshot(json)
      })
   })
   describe('composite doc', () => {
      test('many circles', () => {
         const circle1 = toSvg.circle({
            center: {x: 500, y: 500},
            r: 500,
            fill: 'yellow',
            stroke: 'none'
         })
         const circle2 = toSvg.circle({
            center: {x: 500, y: 500},
            r: 400,
            fill: 'green',
            stroke: 'none'
         })
         const circle3 = toSvg.circle({
            center: {x: 500, y: 500},
            r: 300,
            fill: 'blue',
            stroke: 'none'
         })
         expectSnapshot([
            circle1,
            circle2,
            circle3
         ])
      })
   })
})

