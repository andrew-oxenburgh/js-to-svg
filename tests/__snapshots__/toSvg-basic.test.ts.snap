// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animateMotion 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>animateMotion</title>
  <g>
    <rect fill="#8FBC8F" stroke="none" x="0" y="0" width="200" height="100"/>
    <path fill="none" stroke="lightgrey" d="M20,50 C20,-50 180,150 180,50 C180-50 20,150 20,50 z"/>
    <circle fill="coral" r="5">
      <animateMotion dur="10s" repeatCount="indefinite" path="M20,50 C20,-50 180,150 180,50 C180-50 20,150 20,50 z"/>
    </circle>
  </g>
</svg>"
`;

exports[`animation circle radius 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>a bunch of really ugly animations</title>
  <rect fill="red" x="0" y="0" width="2000" height="2000"/>
  <circle fill="red" r="50" stroke="none" cx="500" cy="500">
    <animate attributeName="r" values="0;100;0" dur="5s" repeatCount="indefinite"/>
    <animate attributeName="fill" values="blue;red;green" dur="3s" repeatCount="indefinite"/>
    <animate attributeName="cx" values="0;400" dur="5s" repeatCount="indefinite"/>
    <animate attributeName="cy" values="0;400" dur="5s" repeatCount="indefinite"/>
  </circle>
</svg>"
`;

exports[`animation rect rx 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a rect given a size and an origin</title>
  <rect fill="red" stroke="none" width="200" height="200" x="0" y="0">
    <animate attributeName="rx" values="0;200;0" dur="2s" repeatCount="indefinite"/>
  </rect>
</svg>"
`;

exports[`boxesWithinBoxes 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>boxesWithinBoxes</title>
  <g>
    <rect fill="green" stroke="none" x="0" y="0" width="1000" height="1000"/>
    <rect fill="orange" stroke="none" x="100" y="100" width="800" height="800"/>
    <rect fill="yellow" stroke="none" x="200" y="200" width="600" height="600"/>
    <rect fill="purple" stroke="none" x="300" y="300" width="400" height="400"/>
  </g>
</svg>"
`;

exports[`clipPath heart 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>demonstrate clipping</title>
  <clipPath id="myClip">
    <circle fill="red" cx="40" cy="40" r="20"/>
  </clipPath>
  <path id="heart" d="M10,30 A20,20,0,0,1,50,30 A20,20,0,0,1,90,30 Q90,60,50,90 Q10,60,10,30 Z"/>
  <use clip-path="url(#myClip)" href="#heart" fill="red"/>
</svg>"
`;

exports[`svg circle bounding square 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a circle with a bounding square</title>
  <circle fill="red" stroke="none" cx="250" cy="250" r="250"/>
</svg>"
`;

exports[`svg circle point and radius 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a circle with a center point and a radius</title>
  <circle fill="red" r="233" stroke="none" cx="181" cy="191"/>
</svg>"
`;

exports[`svg circle standard 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a circle with standard svg attributes</title>
  <circle fill="yellow" cx="500" cy="500" r="300" stroke="none"/>
</svg>"
`;

exports[`svg composite doc many circles 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>got lots of things going on</title>
  <circle fill="yellow" r="500" stroke="none" cx="500" cy="500"/>
  <circle fill="green" r="400" stroke="none" cx="500" cy="500"/>
  <circle fill="blue" r="300" stroke="none" cx="500" cy="500"/>
  <circle fill="red" stroke="none" cx="500" cy="500" r="200"/>
</svg>"
`;

exports[`svg doc empty doc 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create an empty doc</title>
</svg>"
`;

exports[`svg line a short line 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a path</title>
  <line stroke-linecap="round" fill="red" stroke="orange" stroke-width="10" x1="300" y1="350" x2="900" y2="500"/>
</svg>"
`;

exports[`svg path a short line 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a path</title>
  <path d="M 100 100 H 300 Z" fill="red" stroke="orange" stroke-width="10"/>
</svg>"
`;

exports[`svg rect origin and size 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a rect with an origin and a size</title>
  <rect fill="red" stroke="none" width="345" height="678" x="0" y="1"/>
</svg>"
`;

exports[`svg rect rect 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a rect with a input rect structure</title>
  <rect fill="red" stroke="none" x="1" y="2" width="100" height="200"/>
</svg>"
`;

exports[`svg rect standard 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a rect using standard svg attributes</title>
  <rect width="777" height="888" x="0" y="1" fill="red" stroke="none"/>
</svg>"
`;

exports[`svg text a word 1`] = `
"<svg 
  xmlns="http://www.w3.org/2000/svg" width="1000" height="1000">
  <title>create a path</title>
  <text stroke-width="4" stroke="red" fill="none" font=" 130px italic" x="0" y="500">some cool text</text>
</svg>"
`;
